feature/#166 _4 - put back pantries and shopping lists, as regular "database" product lists

monsieurtanuki opened the pull request:
Now the database and the code are much more consistent.
Is that relevant to display pantries and shopping lists in the home page? That's another story.
But at least for the tests and for brainstorming purposes, it's good to see that it's easy to create product list types with different UIs.

Deleted files:
* `pantry.dart`
* `pantry_button.dart`
* `pantry_dialog_helper.dart`
* `pantry_list_page.dart`
* `pantry_page.dart`
* `pantry_preview.dart`

New files:
* `product_list_add_button.dart`: "Add button" related to a user product list type
* `product_list_item.dart`: Widget for a `ProductList` item, for all product list types
* `product_list_item_pantry.dart`: Widget for a `ProductList` item (pantry)
* `product_list_item_shopping.dart`: Widget for a `ProductList` item (shopping list)
* `product_list_item_simple.dart`: Widget for a `ProductList` item (simple product list)

Impacted files:
* `home_page.dart`: now displays pantries and shopping lists (as product lists) and removed dead commented code
* `list_page.dart`: now manages new product list types; now uses new class `ProductListAddButton`
* `multi_select_product_page.dart`: simplified by removing the pantries and shopping lists, as they are now just special product lists
* `product_copy_helper.dart`: simplified by removing the pantries and shopping lists, as they are now just special product lists
* `product_copy_view.dart`: simplified by removing the pantries and shopping lists, as they are now just special product lists
* `product_list.dart`: create new list types for pantries and shopping lists;
* `product_list_button.dart`: moved code to new class `ProductListAddButton` - here we only care about
* `product_list_dialog_helper.dart`: now manages new product list types, more specifically in method `openNew`
* `product_list_page.dart`: now manages new product list types, by using new class `ProductListItem`
* `product_page.dart`: simplified by removing the pantries and shopping lists, as they are now just special product lists
* `product_query_page_helper.dart`: handled the new product list types; refactored
* `smooth_product_card_found.dart`: unrelated fix
* `user_preferences.dart`: removed everything related to pantry and shopping

monsieurtanuki committed:
feature/#166 _4 - put back pantries and shopping lists, as regular "database" product lists

Now the database and the code is much more consistent.
Is that relevant to display pantries and shopping lists in the home page? That's another story.
But at least for the tests and for brainstorming purposes, it's good to see that it's easy to create product list types with different UIs.

Deleted files:
* `pantry.dart`
* `pantry_button.dart`
* `pantry_dialog_helper.dart`
* `pantry_list_page.dart`
* `pantry_page.dart`
* `pantry_preview.dart`

New files:
* `product_list_add_button.dart`: "Add button" related to a user product list type
* `product_list_item.dart`: Widget for a `ProductList` item, for all product list types
* `product_list_item_pantry.dart`: Widget for a `ProductList` item (pantry)
* `product_list_item_shopping.dart`: Widget for a `ProductList` item (shopping list)
* `product_list_item_simple.dart`: Widget for a `ProductList` item (simple product list)

Impacted files:
* `home_page.dart`: now displays pantries and shopping lists (as product lists) and removed dead commented code
* `list_page.dart`: now manages new product list types; now uses new class `ProductListAddButton`
* `multi_select_product_page.dart`: simplified by removing the pantries and shopping lists, as they are now just special product lists
* `product_copy_helper.dart`: simplified by removing the pantries and shopping lists, as they are now just special product lists
* `product_copy_view.dart`: simplified by removing the pantries and shopping lists, as they are now just special product lists
* `product_list.dart`: create new list types for pantries and shopping lists;
* `product_list_button.dart`: moved code to new class `ProductListAddButton` - here we only care about
* `product_list_dialog_helper.dart`: now manages new product list types, more specifically in method `openNew`
* `product_list_page.dart`: now manages new product list types, by using new class `ProductListItem`
* `product_page.dart`: simplified by removing the pantries and shopping lists, as they are now just special product lists
* `product_query_page_helper.dart`: handled the new product list types; refactored
* `smooth_product_card_found.dart`: unrelated fix
* `user_preferences.dart`: removed everything related to pantry and shopping

M123-dev commented on packages/smooth_app/lib/bottom_sheet_views/product_copy_view.dart:
```diff
@@ -1,92 +1,73 @@
 
   @override
   Widget build(BuildContext context) {
+    final List<Widget> items = <Widget>[];
+    items.add(
+      Container(
+        padding: const EdgeInsets.symmetric(horizontal: 20.0),
+        margin: const EdgeInsets.only(top: 20.0, bottom: 24.0),
+        child: Text(
+          'Add this product',
+          style: Theme.of(context).textTheme.headline1,
+        ),
+      ),
+    );
+    for (final String productListType in _USER_PRODUCT_TYPES) {
+      items.add(
+        Container(
+          alignment: Alignment.center,
+          child: Text(
+            _getLabel(productListType),
+            style: Theme.of(context).textTheme.headline2,
+          ),
+          width: MediaQuery.of(context).size.width,
+        ),
+      );
+      items.add(
+        Container(
+          margin: const EdgeInsets.symmetric(horizontal: 5),
+          child: Wrap(
+            direction: Axis.horizontal,
+            children: children[productListType]!,
+            spacing: 8.0,
+          ),
+        ),
+      );
+    }
+    items.add(
+      SizedBox(height: MediaQuery.of(context).size.height * 0.05),
+    );
```
Why do you add all widgets individually to the list?

monsieurtanuki replied:
@M123-dev There's a loop. Of course I could also pre-compute the list length and use `ListView.builder`, but the code wouldn't be easier to read, would it?

M123-dev commented on packages/smooth_app/lib/pages/product/common/product_list_add_button.dart:
```diff
@@ -0,0 +1,30 @@
+
+  final Function onPressed;
+  final bool onlyIcon;
+  final String productListType;
```
A general question, wouldn't it be easier to extract the correct enum value from the string as early as possible and then pass this on?

monsieurtanuki replied:
@M123-dev At some point we need `String`s because we store the product list type in the database.
`enum` would make sense, of course, but we'll still have to convert to/from `String`.
IMHO no rush for creating `enum`s.

M123-dev commented on packages/smooth_app/lib/pages/product/common/product_list_item_shopping.dart:
```diff
@@ -0,0 +1,95 @@
+        reorderIndex: reorderIndex,
+      ),
+      const Divider(color: Colors.grey),
+      Row(
+        mainAxisAlignment: MainAxisAlignment.spaceBetween,
+        children: <Widget>[
+          Row(
+            mainAxisAlignment: MainAxisAlignment.start,
+            children: <Widget>[
+              IconButton(
+                onPressed: () async => await _add(-1),
+                icon: const Icon(Icons.remove_circle_outline),
+              ),
+              Text('${_getCount()}', style: const TextStyle(fontSize: 16)),
+              IconButton(
+                onPressed: () async => await _add(1),
+                icon: const Icon(Icons.add_circle_outline),
+              ),
+            ],
+          ),
+        ],
+      ),
+    ]);
```
Here we only need one row

monsieurtanuki replied:
@M123-dev Fair enough.

M123-dev reviewed (approved):
Thank you @monsieurtanuki, please have a look at my comments, nothing important therefore I already appove

monsieurtanuki committed:
feature/#166 _4 - put back pantries and shopping lists, as regular "database" product lists

Now the database and the code is much more consistent.
Is that relevant to display pantries and shopping lists in the home page? That's another story.
But at least for the tests and for brainstorming purposes, it's good to see that it's easy to create product list types with different UIs.

Deleted files:
* `pantry.dart`
* `pantry_button.dart`
* `pantry_dialog_helper.dart`
* `pantry_list_page.dart`
* `pantry_page.dart`
* `pantry_preview.dart`

New files:
* `product_list_add_button.dart`: "Add button" related to a user product list type
* `product_list_item.dart`: Widget for a `ProductList` item, for all product list types
* `product_list_item_pantry.dart`: Widget for a `ProductList` item (pantry)
* `product_list_item_shopping.dart`: Widget for a `ProductList` item (shopping list)
* `product_list_item_simple.dart`: Widget for a `ProductList` item (simple product list)

Impacted files:
* `home_page.dart`: now displays pantries and shopping lists (as product lists) and removed dead commented code
* `list_page.dart`: now manages new product list types; now uses new class `ProductListAddButton`
* `multi_select_product_page.dart`: simplified by removing the pantries and shopping lists, as they are now just special product lists
* `product_copy_helper.dart`: simplified by removing the pantries and shopping lists, as they are now just special product lists
* `product_copy_view.dart`: simplified by removing the pantries and shopping lists, as they are now just special product lists
* `product_list.dart`: create new list types for pantries and shopping lists;
* `product_list_button.dart`: moved code to new class `ProductListAddButton` - here we only care about
* `product_list_dialog_helper.dart`: now manages new product list types, more specifically in method `openNew`
* `product_list_page.dart`: now manages new product list types, by using new class `ProductListItem`
* `product_page.dart`: simplified by removing the pantries and shopping lists, as they are now just special product lists
* `product_query_page_helper.dart`: handled the new product list types; refactored
* `smooth_product_card_found.dart`: unrelated fix
* `user_preferences.dart`: removed everything related to pantry and shopping

monsieurtanuki commented:
Thank you @M123-dev for your code review!
